lib/model/message.dart

```diff
@@ -180,6 +183,64 @@ class MessageStoreImpl extends PerAccountStoreBase with MessageStore, _OutboxMes
     _disposed = true;
   }
 
+  static const _markReadOnScrollBatchSize = 1000;
+  static const _markReadOnScrollDebounceDuration = Duration(milliseconds: 500);
+  final _markReadOnScrollQueue = _MarkReadOnScrollQueue();
+  bool _markReadOnScrollBusy = false;
+
+  /// Returns true on success, false on failure.
+  Future<bool> _sendMarkReadOnScrollRequest(Iterable<int> toSend) async {
+    assert(toSend.isNotEmpty);
+
+    // TODO(#1581) mark as read locally for latency compensation
+    //   (in Unreads and on the message objects)
+    try {
+      await updateMessageFlags(connection,
+        messages: toSend.toList(),
+        op: UpdateMessageFlagsOp.add,
+        flag: MessageFlag.read);
+    } on ApiRequestException {
+      // TODO(#1581) un-mark as read locally?
+      return false;
+    }
+    return true;
+  }
+
+  @override
+  void markReadOnScroll(Iterable<int> messageIds) async {
+    assert(!_disposed);
+    final anyAdded = _markReadOnScrollQueue.add(messageIds);
+    if (!anyAdded) return;
+    if (_markReadOnScrollBusy) return;
+
+    _markReadOnScrollBusy = true;
+    do {
+      final toSend = <int>[];
+      int i = 0;
+      for (final messageId in _markReadOnScrollQueue.iterable) {
+        if (toSend.length == _markReadOnScrollBatchSize) {
+          break;
+        }
+        i++;
```

Comment:
nit: a more descriptive name is probably helpful for this; perhaps `numFromQueue`?